common/src/main/java/com/viaversion/viaversion/protocols/protocol1_13to1_12_2/packets/WorldPackets.java

```diff
@@ -334,77 +336,63 @@ public void registerMap() {
                                 }
                             }
                         }
 
                         save_connections:
                         {
-                            if (!Via.getConfig().isServersideBlockConnections()
-                                    || !ConnectionData.needStoreBlocks()) break save_connections;
+                            if (!Via.getConfig().isServersideBlockConnections() || !ConnectionData.needStoreBlocks()) break save_connections;
 
                             if (!chunk.isFullChunk()) { // Update
-                                ConnectionData.blockConnectionProvider.unloadChunkSection(wrapper.user(), chunk.getX(), i, chunk.getZ());
+                                ConnectionData.blockConnectionProvider.unloadChunkSection(wrapper.user(), chunk.getX(), s, chunk.getZ());
                             }
                             boolean willSave = false;
-                            for (int j = 0; j < section.getPaletteSize(); j++) {
-                                if (ConnectionData.isWelcome(section.getPaletteEntry(j))) {
+                            for (int p = 0; p < blocks.size(); p++) {
+                                if (ConnectionData.isWelcome(blocks.idByIndex(p))) {
                                     willSave = true;
                                     break;
                                 }
                             }
                             if (!willSave) break save_connections;
 
-                            for (int y = 0; y < 16; y++) {
-                                for (int z = 0; z < 16; z++) {
-                                    for (int x = 0; x < 16; x++) {
-                                        int block = section.getFlatBlock(x, y, z);
-                                        if (ConnectionData.isWelcome(block)) {
-                                            int globalX = x + (chunk.getX() << 4);
-                                            int globalY = y + (i << 4);
-                                            int globalZ = z + (chunk.getZ() << 4);
-                                            ConnectionData.blockConnectionProvider.storeBlock(wrapper.user(), globalX,
-                                                    globalY, globalZ, block);
-                                        }
-                                    }
+                            for (int idx = 0; idx < ChunkSection.SIZE; idx++) {
+                                int id = blocks.idAt(idx);
+                                if (ConnectionData.isWelcome(id)) {
+                                    int globalX = (idx & 0xF) + (chunk.getX() << 4);
+                                    int globalY = (idx >> 8 & 0xF) + (s << 4);
+                                    int globalZ = (idx >> 4 & 0xF) + (chunk.getZ() << 4);
```

Comment:
Should probably add static `xFromIndex`, y, z methods in ChunkSection so these kinds of lines are easier to understand